git/git.go

```diff
@@ -0,0 +1,167 @@
+	When  time.Time
+}
+
+// Commit contains all possible information about a Git commit.
+type Commit struct {
+	// Hash is the SHA1 hash of the commit.
+	Hash Hash
+	// Reference is the original reference of the commit, for example:
+	// 'refs/tags/foo'.
+	Reference string
+	// Author is the original author of the commit.
+	Author Signature
+	// Committer is the one performing the commit, might be different from
+	// Author.
+	Committer Signature
+	// Signature is the PGP signature of the commit.
+	Signature string
+	// Encoded is the encoded commit, without any signature.
+	Encoded []byte
+	// Message is the commit message, contains arbitrary text.
+	Message string
+}
+
+// String returns a string representation of the Commit, composed
+// out the last part of the Reference element, and/or Hash.
+// For example: 'tag-1/a0c14dc8580a23f79bc654faa79c4f62b46c2c22',
+// for a "tag-1" tag.
+func (c *Commit) String() string {
+	if short := strings.SplitAfterN(c.Reference, "/", 3); len(short) == 3 {
+		return fmt.Sprintf("%s/%s", short[2], c.Hash)
+	}
+	return fmt.Sprintf("HEAD/%s", c.Hash)
+}
+
+// Verify the Signature of the commit with the given key rings.
+// It returns the fingerprint of the key the signature was verified
+// with, or an error.
+func (c *Commit) Verify(keyRing ...string) (string, error) {
+	if c.Signature == "" {
+		return "", fmt.Errorf("commit does not have a PGP signature")
+	}
+
+	for _, r := range keyRing {
+		reader := strings.NewReader(r)
+		keyring, err := openpgp.ReadArmoredKeyRing(reader)
+		if err != nil {
+			return "", fmt.Errorf("unable to read armored key ring: %w", err)
+		}
+		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, bytes.NewBuffer(c.Encoded), bytes.NewBufferString(c.Signature), nil)
+		if err == nil && len(signer.PrimaryKey.Fingerprint) >= 20 {
+			return fmt.Sprintf("%X", signer.PrimaryKey.Fingerprint[12:20]), nil
+		}
+	}
+	return "", fmt.Errorf("unable to verify commit with any of the given key rings")
+}
+
+// ShortMessage returns the first 50 characters of a commit subject.
+func (c *Commit) ShortMessage() string {
+	subject := strings.Split(c.Message, "\n")[0]
+	r := []rune(subject)
+	if len(r) > 50 {
+		return fmt.Sprintf("%s...", string(r[0:50]))
+	}
+	return subject
+}
+
+// GitClient defines all the methods that a git implementation must cover.
+// An instance of GitClient should have a 1:1 mapping with a Git repository.
+type GitClient interface {
+	// Init initializes a repository at the configured path with the origin
+	// remote set to url on the provided branch.
+	Init(ctx context.Context, url, branch string) error
+	// Clone clones a repository from the provided url using the options provided.
+	Clone(ctx context.Context, url string, checkoutOpts CheckoutOptions) (*Commit, error)
+	// Push pushes the current branch of the repository to origin.
+	Push(ctx context.Context) error
+	// SwitchBranch switches the active branch of the repository to the
+	// provided branch.
```

Comment:
The `go-git` implementation creates a new branch if one does not exist. Please expand the documentation to that effect.